src/allmydata/client.py

```diff
@@ -582,6 +639,136 @@ def init_storage(self):
+        self._enable_storage_servers(announceable_storage_servers)
+
+
+    def _get_enabled_storage_plugin_names(self):
+        """
+        Get the names of storage plugins that are enabled in the configuration.
+        """
+        return set(
+            self.config.get_config(
+                "storage", "plugins", b""
+            ).decode("ascii").split(u",")
+        )
+
+
+    def _collect_storage_plugins(self, storage_plugin_names):
+        """
+        Get the storage plugins with names matching those given.
+        """
+        return list(
+            plugin
+            for plugin
+            in getPlugins(IFoolscapStoragePlugin)
+            if plugin.name in storage_plugin_names
+        )
+
+
+    def _create_plugin_storage_servers(self, plugins):
+        """
+        Cause each storage plugin to instantiate its storage server and return
+        them all.
+
+        :return: A ``Deferred`` that fires with storage servers instantiated
+            by all of the given storage server plugins.
+        """
+        return defer.gatherResults(
+            list(
+                plugin.get_storage_server(
+                    self._get_storage_plugin_configuration(plugin.name),
+                    lambda: self.getServiceNamed(StorageServer.name),
+                ).addCallback(
+                    partial(
+                        self._add_to_announcement,
+                        {u"name": plugin.name},
+                    ),
+                )
+                for plugin
+                # The order is fairly arbitrary and it is not meant to convey
+                # anything but providing *some* stable ordering makes the data
+                # a little easier to deal with (mainly in tests and when
+                # manually inspecting it).
+                in sorted(plugins, key=lambda p: p.name)
+            ),
+        )
+
+
+    def _add_to_announcement(self, information, announceable_storage_server):
+        """
+        Create a new ``AnnounceableStorageServer`` based on
+        ``announceable_storage_server`` with ``information`` added to its
+        ``announcement``.
+        """
+        updated_announcement = announceable_storage_server.announcement.copy()
+        updated_announcement.update(information)
+        return AnnounceableStorageServer(
+            updated_announcement,
+            announceable_storage_server.storage_server,
+        )
+
+
+    def _get_storage_plugin_configuration(self, storage_plugin_name):
+        """
+        Load the configuration for a storage server plugin with the given name.
+
+        :return dict: The matching configuration.
+        """
+        try:
+            # Need to reach past the Tahoe-LAFS-supplied wrapper around the
+            # underlying ConfigParser...
+            config = self.config.config.items(
+                "storageserver.plugins." + storage_plugin_name,
+            )
+        except NoSectionError:
```

Comment:
I feel like the Tahoe-LAFS wrapper should provide everything we need internally; I know in some branch I introduced `enumerate_section` or similar? (Possibly in grid-manager? sorry don't have it at hand right now).

Anyway, most of the point of the `_Config` thing is so that *some* day we could maybe use a thing besides "a config file" and/or have other ways of getting/supplying config information .. Currently a lot of the API there is fairly tied to what config-files do, but that was/is the goal.